### PR TITLE
Switch SVD on CPU from gesvd to gesdd

### PR DESCRIPTION
**Pull request by vishwakftw**
- Added a note to the doc string for `svd`.

### CODE REVIEW

**Comment by vishwakftw** [comment]
Will it be possible to get some benchmark info on this? I've been fixing various issues on my laptop after upgrading to Ubuntu 18.04.

**Comment by zou3519** [comment]
The performance speedup is great, @vishwakftw! Could you benchmark some smaller matrices? In particular, I'm reading https://groups.google.com/forum/#!topic/julia-dev/mmgO65i6-fA and from their experience it sounds like there is some value in having a second svd call that uses gesvd, but I'm curious what your benchmarking shows about that.

File: torch/_torch_docs.py [hunk 1]
```diff
@@ -4409,6 +4409,10 @@ def parse_kwargs(desc):
 If :attr:`some` is ``True`` (default), the returned `U` and `V` matrices will
 contain only :math:`min(n, m)` orthonormal columns.
 
+.. note:: The implementation of SVD on CPU uses the LAPACK routine `?gesdd` (a divide-and-conquer
+          algorithm) instead of `?gesvd` for speed. However, the SVD on GPU uses the MAGMA routine
+          `gesvd` since it takes up lesser workspace area.
```
Comment (SsnL) [inline torch/_torch_docs.py]: nit: `?gesvd`

**Review by ssnl — APPROVED** [review]
looking good except for 1 nit

File: torch/_torch_docs.py (hunk 1 above)
Comment (vishwakftw) [reply torch/_torch_docs.py]: I didn't add the `?` considering it is a MAGMA routine that was being discussed there.

**Comment by vishwakftw** [comment]
I have got some more times with me (10 runs w/ 1000 loops each)

| dimensions | before this patch | after this patch |
| --------- | ----------- | ---------- |
| 200 x 100 | 3.3 ms +/- 31.7 us per loop | 2.09 ms +/- 190 us per loop |
| 100 x 50 | 1.01 ms +/- 1.92 us per loop | 634 us +/- 10.9 us per loop | 
| 50 x 25 | 212 us +/- 10.6 us per loop | 215 us +/- 1.8 us per loop |
| 40 x 20 | 126 us +/- 1.53 us per loop | 115 us +/- 2.37 us per loop |
| 20 x 10 | 46.8 us +/- 87.1 ns per loop | 53.5 us +/- 7.91 us per loop |
| 10 x 5 | 28.3 us +/- 395 ns per loop | 33.9 us +/- 8.17 us per loop |

cc: @zou3519 

From this, as the matrix size starts diminishing, the advantage fades away.

**Comment by zou3519** [comment]
@vishwakftw those times look great

File: torch/_torch_docs.py (hunk 1 above)
Comment (vishwakftw) [reply torch/_torch_docs.py]: I just realized that GPUs were running on `gesdd` instead of `gesvd`

**Comment by yaroslavvb** [comment]
BTW, I would also check if the result has any NaN's. I found Intel's gesdd [erroneously produced](https://software.intel.com/en-us/forums/intel-distribution-for-python/topic/628049) NaNs when gesvd didn't.

I feel like some of the edge cases are likely to come up on more ill-conditioned matrices are more likely to break things. For instance here's a particular (very ill-conditioned) matrix that [breaks](https://github.com/tensorflow/tensorflow/issues/9234) TensorFlow's SVD routine

```
from scipy import linalg  # for svd
import urllib.request
import numpy as np

url="https://storage.googleapis.com/tensorflow-community-wheels/svd_in"
response = urllib.request.urlopen(url)
body = response.read()
print("Read %d bytes"%(len(body),))
assert len(body) == 15366400
open("svd_in", "wb").write(body)

dtype = np.float32
matrix0 = np.genfromtxt('svd_in',
                        delimiter= ",").astype(dtype)
assert matrix0.shape == (784, 784)
u, s, v = linalg.svd(matrix0)
print("matrix0 any NaNs: %s"% (np.isnan(matrix0).any(),))
print("u had NaNs: %s"% (np.isnan(u).any(),))
```

**Comment by vishwakftw** [comment]
@SsnL is this good to go?

**Comment by vishwakftw** [comment]
@yaroslavvb The example you have given seems to not break the current implementation in this PR. There were no `NaN`s or `inf`s detected.

**Review by ssnl — APPROVED** [review]

File: torch/_torch_docs.py [hunk 2]
```diff
@@ -4409,6 +4409,10 @@ def parse_kwargs(desc):
 If :attr:`some` is ``True`` (default), the returned `U` and `V` matrices will
 contain only :math:`min(n, m)` orthonormal columns.
 
+.. note:: The implementation of SVD on CPU uses the LAPACK routine `?gesdd` (a divide-and-conquer
```
Comment (SsnL) [inline torch/_torch_docs.py]: Oh so we are running gesdd in both paths. Given this case, could you rename the function name from `gesvd` to `gesdd`?

Comment (vishwakftw) [reply torch/_torch_docs.py]: I was going to ask you about that: shall I proceed to rename all `gesvd` functions to `gesdd` in ATen? I think it should be a simple codemod.

Comment (SsnL) [reply torch/_torch_docs.py]: Yes please! I think only TH, THC, and a Declaraions.cwrap change is needed, so it's not too bad. Thanks!

Comment (vishwakftw) [reply torch/_torch_docs.py]: @SsnL done!!

**Review by soumith — APPROVED** [review]

**Review by soumith — APPROVED** [review]

**Pull request closed without merging**